### ISSUE TITLE
Amend usages of various (sub)commands

### ISSUE THREAD

**Issue by periperidip** [gitgitgadget/git#636]
The Git suite option parsing API's Technical Documentation suggests [this](https://github.com/git/git/blob/v2.26.2/Documentation/technical/api-parse-options.txt#L267) about the option descriptions of a (sub)command:

> `description` is a short string to describe the effect of the option.
  It shall begin with a lower-case letter and a full stop (`.`) shall be
  omitted at the end.

A bit more insight about what exactly are the option descriptions (or rather what options are): Generally, a (sub)command has various options that may go with it, for instance, a popular one may be `quiet` (`--quiet` or `-q`) which suppresses the output of an action performed by a (sub)command; another one may be `verbose` (`--verbose` or `-v`) which gives a bit more detail about the actions being performed by the (sub)command. The job of the `parse_options` API is to intake the arguments passed down via the command line and sieve out the options (if provided) and the non-option arguments (such as a `branch` or a `filename`), thus modifying the final `argv[]` array to hold the non-option arguments leaving out the other arguments (such as the (sub)comand name, unless explicitly specified by the `PARSE_OPT_KEEP_ARGV0` flag).

Whenever the user uses the (sub)command in a wrong manner (such as not providing a mandatory argument), the program exits with an error code prompting out a `usage` which kinda looks like:

```
usage: git submodule--helper set-branch [--quiet] (-d|--default) <path>
or: git submodule--helper set-branch [--quiet] (-b|--branch) <branch> <path>

-q, --quiet           suppress output for setting default tracking branch of a submodule
--default             set the default tracking branch to master
--branch <branch>     set the default tracking branch to the one specified
```

As one may observe above, there are two ways to use the subcommand `set-branch` of `git submodule` along with the options available to use along with it as well as their **description**. The `options` array of the above case looks like:

```
struct option options[] = {
		OPT__QUIET(&quiet, N_("suppress output for setting default tracking branch of a submodule")),
		OPT_BOOL(0, "default", &opt_default, N_("set the default tracking branch to master")),
		OPT_STRING(0, "branch", &opt_branch, N_("branch"), N_("set the default tracking branch to the one specified")),
		OPT_END()
	};
```

Now one can clearly observe where the description is above. Various (sub)commands' (such as [here](https://github.com/git/git/blob/v2.26.2/builtin/submodule--helper.c#L564)) `option` arrays don't follow the guideline provided by the `parse_options` Documentation regarding the descriptions. This needs to be amended.

**Comment by ffyuanda** [comment]
Think this issue should be closed? Cannot find anything alike in the code right now.

**Comment by dscho** [comment]
> Think this issue should be closed? Cannot find anything alike in the code right now.

@ffyuanda You are right in that [the current `git submodule foreach` options](https://github.com/git/git/blob/v2.35.0-rc2/builtin/submodule--helper.c#L560-L565) seem to have been fixed relative to [the version mentioned in the original report](https://github.com/git/git/blob/v2.26.2/builtin/submodule--helper.c#L564).

However, I get this:

```console
$ rev=v2.35.0-rc2 &&
  for entry in $(git grep -l 'struct option .* = {$' $rev -- \*.c); \
  do \
    git show "$entry" |
    sed -n '/struct option .* = {/,/OPT_END/{=;p}' |
    sed "N;s/^\\([0-9]*\\)\\n/$(echo "$entry" | sed 's/\//\\&/g'):\\1/"; \
  done |
  grep -Pe '((?<!OPT_GROUP\(N_\(|OPT_GROUP\()"(?!GPG|DEPRECATED|SHA1|HEAD)[A-Z]|(?<!"|\.\.)\.")'

v2.35.0-rc2:builtin/bisect--helper.c:1212                        N_("use <cmd>... to automatically bisect."), BISECT_RUN),
v2.35.0-rc2:builtin/submodule--helper.c:1877            OPT__QUIET(&quiet, "Suppress output for cloning a submodule"),
v2.35.0-rc2:diff.c:5599           N_("Output to a specific file"),
v2.35.0-rc2:t/helper/test-run-command.c:224             OPT_INTEGER('n', "trials", &trials, "Number of trials"),
v2.35.0-rc2:t/helper/test-run-command.c:225             OPT_INTEGER('s', "skip", &skip, "Skip <n> trials"),
v2.35.0-rc2:t/helper/test-run-command.c:226             OPT_BOOL('m', "msys2", &msys2, "Test quoting for MSYS2's sh"),
```

This is substantially reduced compared to running the command with `rev=v2.26.2`, but it is not empty.

Note: The command shown above (with slight modifications) would probably make for a fine addition to `.github/workflows/main.yml` in the same patch series that fixes the remaining issues.

**Comment by Abhra303** [comment]
Hi @dscho ,

> The command shown above (with slight modifications) would probably make for a fine addition to .github/workflows/main.yml in the same patch series that fixes the remaining issues.

Where should I include the check in main.yml? I am little bit confused thinking about which one is the ideal `job` where I can place this check as a `step`. Is `regular` ideal for this check?

**Comment by dscho** [comment]
> Where should I include the check in main.yml?

I'd go for the `documentation` job, namely [`ci/testdocumentation.sh`](https://github.com/git/git/blob/v2.35.1/ci/test-documentation.sh), probably best as a sister target to [`check-builtins`](https://github.com/git/git/blob/v2.35.1/Makefile#L3379-L3382) (maybe `check-usage-strings.sh`?).

**Comment by dscho** [comment]
Oh, and you will want to extend the `git grep -l 'struct option .* = {$' $rev -- \*.c` call to handle cases where the Makefile target/script is run on code extracted from a `.tar` file, probably guarding [the suggested code](https://github.com/gitgitgadget/git/issues/636#issuecomment-1018660439) in a `if test -d .git` conditional block.

**Comment by dscho** [comment]
> > probably best as a sister target to [check-builtins](https://github.com/git/git/blob/v2.35.1/Makefile#L3379-L3382) (maybe check-usage-strings.sh?).
> 
> But there are some files (like `t/helper/test-run-command.c`, `diff.c` etc.) that are not part of `builtin` but they use `struct option` (see the output of your [suggested command](https://github.com/gitgitgadget/git/issues/636#issuecomment-1018660439)) . Would it be good to add that check as a sister target to `check-builtins`?

By "sister target" I meant a completely new target, e.g. `check-usage-strings`, along with the corresponding script, `check-usage-strings.sh`.

Otherwise you'd be correct, "builtins" would be too narrow a scope.

**Comment by Abhra303** [comment]
> By "sister target" I meant a completely new target, e.g. check-usage-strings, along with the corresponding script, check-usage-strings.sh.

Oops. Thanks for the info :)
